src/caduceus_inf.py

```diff
@@ -70,15 +71,60 @@ def generate_soft_labels(
     tokenizer: PreTrainedTokenizer = AutoTokenizer.from_pretrained(
         model_name, trust_remote_code=True
     )
-    model = AutoModelForMaskedLM.from_pretrained(model_name, trust_remote_code=True).to(
-        device
-    )
+
+    # Load the model initially to CPU. This is important for the sequential warm-up.
+    model = AutoModelForMaskedLM.from_pretrained(model_name, trust_remote_code=True)
+
+    if device == "cuda" and torch.cuda.is_available():
+        num_gpus = torch.cuda.device_count()
+        if num_gpus > 1:
+            # === SEQUENTIAL WARM-UP ===
+            # We warm up each GPU sequentially to prevent a race
+            # condition in the Triton kernel autotuner when used with nn.DataParallel.
+            print("Pre-warming Triton cache on all available GPUs sequentially...")
+            for i in range(num_gpus):
+                gpu_device = f"cuda:{i}"
+                print(f"  Warming up on {gpu_device}...")
+                try:
+                    # Move model to the target GPU
+                    model.to(gpu_device)
+
+                    # Create a dummy input on the target GPU
+                    dummy_input = torch.randint(
+                        0,
+                        tokenizer.vocab_size,
+                        (1, chunk_size),
+                        dtype=torch.long,
+                        device=gpu_device,
+                    )
+
+                    # Run a single forward pass to trigger JIT compilation
+                    with torch.inference_mode(), autocast(enabled=True):
+                        _ = model(dummy_input)
+
+                    # Wait for all kernels to finish
+                    torch.cuda.synchronize(gpu_device)
+                    print(f"  Warm-up on {gpu_device} complete.")
+
+                except Exception as e:
+                    print(f"  An error occurred during warm-up on {gpu_device}: {e}")
```

Comment:
Printing traceback here could make this more readable. You can see in #11. But in general this seems like a good use case for `logger.exception` 🤔

Reply:
Great point, I'll add this tomorrow, when I debug the issue with NVidia driver. Thanks!